### ISSUE TITLE
ES6 modules/import support

### ISSUE THREAD

**Issue by josephrocca** [pyodide/pyodide#1477]
Is there any desire from the core team to move to ES6 modules? This would make it easier to work with for modern JS development and it would also make it compatible with [Deno](https://deno.land) out-of-the-box, since Deno is highly compatible with the browser JS environment (though not to the point of supporting script element creation as in the snippet below, hence the current lack of Deno support). 

https://github.com/pyodide/pyodide/blob/093c0dd18ddcefa4cc1bce5f404f4cec2444ef9d/src/pyodide.js#L58-L73

Just now I "wrapped" `ftfy` into [a JS module](https://github.com/josephrocca/ftfy-pyodide) and I am **amazed** at how easy/seamless it was. The Python module is only 4x faster than this Pyodide-based JS one, which is significantly better than I was expecting. I'd love to be able to bring it to the server-side, since there isn't any JS module like this as far as I'm aware.

It's kind of mind blowing to me that, thanks to Pyodide, the JS ecosystem can suddenly benefit from the many, many developer-years that have gone into the Python package ecosystem.

There are quite a few Python packages like `ftfy` where no comparable JS package exists, and so I'd love to be able to bring them the both the browser and the server. Using ES6 modules would make that possible.

I would be happy to do the work required to move to ES6 modules - I'd just need to some guidance on what changes would need to be made.

**Comment by rth** [comment]
Hi @josephrocca,

yes it would be great for 0.18.0. 

Right now we are focused on the 0.17.0 release which will happen by the end of next week. After that we should make a plan on how to incrementally make this transition (see related issues https://github.com/pyodide/pyodide/issues/14 https://github.com/pyodide/pyodide/issues/328 https://github.com/pyodide/pyodide/pull/792 ), and your contribution would certainly be very welcome!

**Comment by josephrocca** [comment]
@rth Okay, great! Let me know when you've got some work for me to do 👍

**Comment by hoodmane** [comment]
@josephrocca Thanks for this suggestion.

> Is there any desire from the core team to move to ES6 modules?

Yes so much desire for this (from me anyways). With this we could eliminate the rest of the global variables from the Pyodide runtime, and also our current loading scheme is really terrible and module loading is 
 See #1364 for my most recent attempt.

Currently we are blocked on Firefox/Safari support for using ES6 Modules in webworkers. It seems that this feature landed in Firefox 8 days ago and in Safari in February (two months ago):
https://bugzilla.mozilla.org/show_bug.cgi?id=1247687
https://bugs.webkit.org/show_bug.cgi?id=164860
whereas chrome has supported the feature for about a year (since version 80), though service workers can only use modules starting in chrome v91.
https://bugs.chromium.org/p/chromium/issues/detail?id=680046
https://bugs.chromium.org/p/chromium/issues/detail?id=824647

Our current CanIUse policy is pretty loose with using recent features, but I think August 2022 is about the earliest we could demand Module support.

So the question is how can I make an object that can either be loaded as an ES6 module or using `importScripts`? We need to be able to feature detect the `export` syntax. The problem is that as far as I can tell anything that uses `export` raises a `SyntaxError` if it is missing.

But yeah, if you could figure out a way to say:
```js
if(loadedAsModule){
   export { installPackage };
} else {
   globalThis.pyodide._module.installPackageFunctions[package_name] = installPackage;
}
```
without a SyntaxError, then we could start using ES6 modules much earlier. Or maybe you could somehow modify the source code in cases when we want a webworker? Perhaps the most promising idea would be to build two copies of `some_package.js`, one as a Module and not as a Module. These will differ by only a few bytes though.

Next question is how to feature detect dynamic import, though this seems to have a solution:
https://stackoverflow.com/questions/60317251/how-to-feature-detect-whether-a-browser-supports-dynamic-es6-module-loading

Anyways if you could get something like #1364 to work that's where I'd start.

**Comment by hoodmane** [comment]
So then the question is how we can use Modules as our loading scheme sometimes and use the old approach other times.

One thing that would be useful would be to make a script that can either be loaded as an ES6 module or using `importScripts`. We need to be able to feature detect the `export` syntax. The problem is that as far as I can tell anything that uses `export` raises a `SyntaxError` if it is missing.

But yeah, if you could figure out a way to say:
```js
if(loadedAsModule){
   export { installPackage };
} else {
   globalThis.pyodide._module.installPackageFunctions[package_name] = installPackage;
}
```
without a SyntaxError, then we could start using ES6 modules much earlier. 

Or maybe you could modify the source code in cases when we want a webworker? Perhaps the most promising idea would be to build two copies of `some_package.js`, one as a Module and not as a Module. These will differ by only a few bytes though.

Next question is how to feature detect dynamic import, though this seems to have a solution:
https://stackoverflow.com/questions/60317251/how-to-feature-detect-whether-a-browser-supports-dynamic-es6-module-loading

I think the best choice here would be to start out by loading the module version on the main thread and in a worker load the old way. That'd get us started quickly.

**Comment by josephrocca** [comment]
@hoodmane It might be easier to just do everything with modern ES6 modules, and then have a simple build step which transpiles to the older-style code with [babel](https://babeljs.io/)? I haven't used babel in several years, so I can't remember how it works, but this article makes it look simple: https://medium.com/@SunnyB/how-to-convert-es6-into-es5-using-babel-1b533d31a169

According to that article it's as simple as:

```bash
npm install -D babel-cli
npm install -D babel-preset-env
echo "{\"presets\":[\"env\"]}" > .babelrc
babel src -d build
```

IIRC you can set specific minimum browser versions that you need to support, and then it will transpile everything to old-style code based on the features available in those browsers.

You can play around with it here: https://babeljs.io/repl

**Comment by hoodmane** [comment]
How well does babel work with emscripten?

**Comment by hoodmane** [comment]
I guess we can add `-s EXPORT_ES6=1` to the build flags and then Emscripten will use ES6 modules. I wonder what the generated code looks like.

**Comment by josephrocca** [comment]
@hoodmane Cool, and then it should be as simple as pointing babel at any `.js` files and it will make copies of those that are compatible with older browsers.

There are instructions for `make` usage, among other things here: https://babeljs.io/setup.html

```
SRC = $(wildcard src/*.js)
LIB = $(SRC:src/%.js=lib/%.js)

lib: $(LIB)
lib/%.js: src/%.js babel.config.json
  mkdir -p $(@D)
  babel $< -o $@
```

And there are some examples of `babel.config.json` here: https://babeljs.io/docs/en/usage

```json
{
  "presets": [
    [
      "@babel/env",
      {
        "targets": {
          "edge": "17",
          "firefox": "60",
          "chrome": "67",
          "safari": "11.1"
        },
        "useBuiltIns": "usage",
        "corejs": "3.6.5"
      }
    ]
  ]
}
```

**Comment by hoodmane** [comment]
I don't think this will be quite that simple. Our package loading code is pretty complicated and delicate, it's lines 58 -- 380 in pyodide.js
https://github.com/pyodide/pyodide/blob/4c315c7c2fb24cdd5c57107094b95190091b08d9/src/pyodide.js#L124
My guess is that lots of things to break when you do this. I would be interested to hear how it works out though.

**Comment by josephrocca** [comment]
@hoodmane I'd need someone to walk me through that code to make a more informed comment, but my first thought is that for loading the actual Python packages, couldn't we just keep it as-is? Because we don't know the specific Python packages that are going to be imported until runtime, right? So we can't use `import foo from "./foo.js"`. We could use dynamic imports: `await import("./foo.js")`, but would there be any major benefit?

From a user perspective, I really just want to be able to import the actual pyodide module as an ES6 module, which would allow me to use it in modern JS contexts, and on the server side. If that were the goal, would there be much more to edit other than [`loadScript` in pyodide.js](https://github.com/pyodide/pyodide/blob/4c315c7c2fb24cdd5c57107094b95190091b08d9/src/pyodide.js#L58) and the scripts that it imports? Seems like `loadScript` could just be replaced with `await import(...)` no matter the context (browser/webworker/deno) if the scripts being imported are ES6 modules?

**Comment by hoodmane** [comment]
> From a user perspective, I really just want to be able to import the actual pyodide module as an ES6 module

Yeah agreed that'd be a huge usability benefit. I'm focusing on the main blocker for this which is the package loading system.

Yeah actually I think we could modify `loadScript`
https://github.com/pyodide/pyodide/blob/4c315c7c2fb24cdd5c57107094b95190091b08d9/src/pyodide.js#L59
to have a third case:
```js
if(we're in a module){
   loadScript = async function(src){
      let mod = await import(src + ".mjs"); // grab module version
      mod.installPackage(Module);
   }
} else if(...){
// other two cases
// ...
} 
```
Then we probably could avoid changing the rest of the package loading code. We'd have to convince webpack / rollup / babel / whatever to leave that particular `import` alone, but it might well do that anyways because it can't figure out what is being imported? 

If we split up `pyodide.js` into multiple source files then babel should be able to put them together fine. I just don't think it helps with our package loading.

**Comment by hoodmane** [comment]
Another issue is that various aspects of our code don't work correctly with `"use strict";` and ES Modules use strict mode by default. One easy test is to check what happens if we add `"use strict";` to `pyodide.js`. This is highly desirable for maintainability in any case.

I think that some of the emscripten-generated code is also not strict mode compatible so that's another thing that we will need to solve.

**Comment by hoodmane** [comment]
> couldn't we just keep it as-is? Because we don't know the specific Python packages that are going to be imported until runtime, right? So we can't use import foo from "./foo.js". We could use dynamic ES6 imports: await import("./foo.js"), but would there be any major benefit?

The problem is that we load the script by adding a `<script></script>` tag to the DOM and then storing `pyodide` in a specific global variable which the loaded script accesses. I'm not sure that it's possible to do at all with Modules. If it is possible, it would destroy a lot of the benefits because there still isn't any encapsulation.

**Comment by josephrocca** [comment]
@hoodmane Ah, right, I see what you mean.

I've been playing around with getting Pyodide working on Deno, and it doesn't actually require many changes. This doesn't fix the whole global variable problem, but I think fixing that would be a minor improvement compared to getting Pyodide running on the server.

Here are the changes:

Change `pyodide._module` to `globalThis.pyodide._module` in files loaded by `loadScript()`, like this:
```js
var Module = typeof globalThis.pyodide._module !== "undefined" ? globalThis.pyodide._module : {};
```
Change `_createPyodideModule` to `globalThis._createPyodideModule` at the top of `pyodide.asm.js` like this:
```js
globalThis._createPyodideModule = (function() {
  var _scriptDir = typeof document !== 'undefined' && document.currentScript ? document.currentScript.src : undefined;
  if (typeof __filename !== 'undefined') _scriptDir = _scriptDir || __filename;
  return (
function(_createPyodideModule) {
  globalThis._createPyodideModule = globalThis._createPyodideModule || {};
```

Add this to top of `pyodide.asm.js` (for some reason these variables are not declared? don't see why they'd need to be global):
```js
let baseName, fpcGOT, dyncallGOT, fpVal, dcVal;
```

The only catch is that I need to run Deno with the `--location` option like so:
```bash
deno run --allow-net --location=http://127.0.0.1:3001 main.js
```
because for some reason which I haven't yet investigated, `pyodide.asm.js` and the files loaded by `loadScript` need access to window.location (which Deno doesn't provide by default because that property doesn't make sense on the server side). Shouldn't it just use the `indexURL` I gave it?

In any case, in case you are interested, `main.js` looks like this:
```js
import ftfy from "./mod.js";
let fixed = ftfy.fix_text("(à¸‡'âŒ£')à¸‡");
console.log(fixed);
```
and `mod.js` looks like this:
```js
// polyfills (not supported in Deno):
import XMLHttpRequest from "https://deno.land/x/xmlhttprequest_deno@v0.0.2/mod.js";
globalThis.XMLHttpRequest = XMLHttpRequest;
globalThis.setImmediate = function(fn, ...args) { setTimeout(fn, 0, ...args) };

import "http://127.0.0.1:3001/pyodide.js";

await loadPyodide({indexURL : "http://127.0.0.1:3001/"});

await pyodide.runPythonAsync(`
  import micropip
  await micropip.install('https://deno.land/x/ftfy_pyodide@v0.1/wcwidth-0.2.5-py2.py3-none-any.whl')
  await micropip.install('https://deno.land/x/ftfy_pyodide@v0.1/ftfy-6.0-py3-none-any.whl')
  import ftfy
`);

export default pyodide.globals.get("ftfy");
```

**Comment by hoodmane** [comment]
> Here are the changes:

You mean these are enough to support strict mode?

> Change pyodide._module to globalThis.pyodide._module in files loaded by loadScript(), like this:

Can be done by changing this line:
https://github.com/pyodide/pyodide/blob/62829f45dbc2a57aed0a4ac5adb612827591d93a/pyodide_build/buildpkg.py#L198

> Change _createPyodideModule to globalThis._createPyodideModule at the top of pyodide.asm.js like this:

I bet we can say `-s EXPORT_NAME="'globalThis._createPyodideModule'"` in the Makefile:
https://github.com/pyodide/pyodide/blob/62829f45dbc2a57aed0a4ac5adb612827591d93a/Makefile#L47

> let baseName, fpcGOT, dyncallGOT, fpVal, dcVal;

For this one we definitely need to patch the output file. 


Of course these are all emscripten bugs, so perhaps we should check if there are issues open about these things.

**Comment by hoodmane** [comment]
> for some reason which I haven't yet investigated, pyodide.asm.js and the files loaded by loadScript need access to window.location (which Deno doesn't provide by default because that property doesn't make sense on the server side). If it's for coordinating package downloads, shouldn't it just use the indexURL I gave it?

Well of course that's all from code generated by emscripten:
```
var PACKAGE_PATH;
if (typeof window === "object") {
  PACKAGE_PATH = window["encodeURIComponent"](
      window.location.pathname.toString().substring(
          0, window.location.pathname.toString().lastIndexOf("/")) +
      "/")
} else if (typeof location !== "undefined") {
  PACKAGE_PATH =
      encodeURIComponent(location.pathname.toString().substring(
                             0, location.pathname.toString().lastIndexOf("/")) +
                         "/")
} else {
  throw "using preloaded data can only be done on a web page or in a web worker"
}
```

**Comment by hoodmane** [comment]
There is one other place where it uses location, but I think it's on a code path where it only happens if it thinks it's in a webworker so you are likely to be safe.

**Comment by hoodmane** [comment]
If we could patch that first code out for `var PACKAGE_PATH = indexURL` it'd be great.

**Comment by hoodmane** [comment]
So we could patch file packager:
https://github.com/emscripten-core/emscripten/blob/d0ff93d2723e57154220723bd77bed540d95d453/tools/file_packager.py#L528

**Comment by josephrocca** [comment]
Oh, I forgot to mention that for the `loadScript` code I also changed this:
```js
 else {
    throw new Error("Cannot determine runtime environment");
 }
```
to this:
```js
 else {
    loadScript = async (url) => {
      let jsText = await fetch(url).then(r => r.text());
      eval(jsText);
    };
  }
```
so Deno can actually load the scripts.

> You mean these are enough to support strict mode?

I think strict mode is only enforced for modules, not for scripts loaded/imported by modules. I haven't added `export` to `pyodide.js`, for example - I'm just importing it and thus loading pyodide as a global variable. So `pyodide.js`, `pyodide.asm.js` and other scripts are still non-modules. They just "export" by adding properties to `globalThis`.

Below is an archive of the edited files in case you wanted to have a play around. The only edits I've made are the ones that I mentioned above (unless I've forgotten something else - but I'm pretty sure that was everything).

[pyodide-working-with-deno.zip](https://github.com/pyodide/pyodide/files/6330605/pyodide-working-with-deno.zip)

If you want to try it on Deno, first [install Deno](https://deno.land/#installation) with the one-liner if you haven't already. If you already have it, then run `deno upgrade` to ensure you're on `v1.9`, then serve the contents of that zip on port 3001. E.g. you can run this in the extracted zip directory:
```bash
deno run --allow-net --allow-read=. https://raw.githubusercontent.com/josephrocca/denoSimpleStatic/master/main.ts --port=3001
```
the `--allow-read=.` part only allows reading from the current directory, so it's safe to run, even if my code were malicious (one of the key value-props of Deno) and then:
```bash
deno run --allow-net --location=http://127.0.0.1:3001 main.js
```

You can debug with `--inspect-brk`:
```bash
deno run --allow-net --location=http://127.0.0.1:3001 --inspect-brk main.js
```
Then just open Chrome dev tools and click [this green icon](https://i.imgur.com/EcGzTcQ.png), and it's just like debugging a browser script. Just mentioning all this in case you are not familiar.

> There is one other place where it uses location, but I think it's on a code path where it only happens if it thinks it's in a webworker so you are likely to be safe.

Deno has webworkers too though (it's a very close copy of the browser environment, so almost all things are the same as in the browser), so ideally all location checks could be removed, or at least put in a try/catch and handling the case where an error is thrown.

I think this is the one you're talking about:
```js
    else if (ENVIRONMENT_IS_WEB || ENVIRONMENT_IS_WORKER) {
      if (ENVIRONMENT_IS_WORKER) {
        scriptDirectory = self.location.href;
      } else if (typeof document !== "undefined" && document.currentScript) {
        scriptDirectory = document.currentScript.src;
      }
```
Deno doesn't have the `document` object (again, doesn't make sense on a server), so it skips both of those if conditions (and there's no final `else`), so `scriptDirectory` doesn't actually end up getting used?
Here's another mention of `window.location`:
```js
      DB_NAME: function () {
        return "EM_FS_" + window.location.pathname;
      },
```
But I put a `console.log` in there and it didn't get executed, so I think it's just that `PACKAGE_PATH` code that's causing the problem.

**Comment by rth** [comment]
I don't know if there is much left to do on this issue. The npm package is getting to a good shape. We haven't tested with deno, but this could probably be a separate issue.

**Comment by xhlulu** [comment]
Commenting here since the title seemed relevant, but i'm having a hard time following the discussion here.

I'm interested in loading modules the ES6 way (e.g. react component libraries) and use them inside the `python` scope. Is that something planned/appropriate to discuss in this issue, or should I open a new one?

**Comment by hoodmane** [comment]
@xhlulu it would be better if you opened a new issue with a slightly more detailed explanation about what you are trying to do.

**Comment by xhlulu** [comment]
Thanks, will do!

**Comment by boisgera** [comment]
AFAICT, @xhlulu would like to have ES6 modules proxied to Python and probably registrable as Python modules :

https://starboard.gg/nb/nIkOYp8

I'd like that too 🙂

**Comment by sntran** [comment]
Just want to chime in that the patch here won't work on Deno Deploy, which does not have dynamic import, and does not allow `eval` to be used.

**Comment by rth** [comment]
Thanks, everyone for this discussion! For deno support I have opened a separate issue https://github.com/pyodide/pyodide/issues/3420 which now works with fewer patches that we should just merge (thanks a lot @josephrocca for investigating). 

> AFAICT, xhlulu would like to have ES6 modules proxied to Python and probably registrable as Python modules:

That sounds to me like a somewhat orthogonal feature request. If it's still relevant please open a separate issue.

I'm going to close this issue,  but please don't hesitate to create more specific issues if you believe there are still things that should be addressed related to the above discussion.
